keps/sig-windows/1981-windows-privileged-container-support/README.md

```diff
@@ -459,87 +463,123 @@ Windows privileged containers will be implemented with [Job Objects](https://doc
-- [Linux SandboxSecurityConfig](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L293)
+
+We will need to add a privileged field to the runtime spec. We can model this after the Linux pod security context and container security context that is a boolean that is set to `true` for privileged containers. References:
+
+- [LinuxSandboxSecurityConfig](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L293)
 - [LinuxSandboxSecurityContext](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L28)
 - [LinuxContainerSecurityContext](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L612)
 
-In windows, the same privileged field would need to be added in the windows pod security context (open issue) and the container security context will need to be augmented with the privileged flag.
+For Windows we are proposing the following updates to CRI-API:
 
-The new WindowsPodSandboxConfig:
-```
-message WindowsPodSandboxConfig {.
-	WindowsSandboxSecurityContext security_context = 1;
+Add WindowsPodSandboxConfig (and it to PodSandboxConfig)
+
+```protobuf
+message WindowsPodSandboxConfig {
+  WindowsSandboxSecurityContext security_context = 1;
 }
 ```
 
-A new WindowsSandboxSecurityContext:
-```
+Add WindowsSandboxSecurityContext:
+
+```protobuf
 message WindowsSandboxSecurityContext {
-	string run_as_user = 1;
-	bool privileged = 2;
+  string run_as_username = 1;
+  bool privileged = 2;
 }
 ```
 
-The new field for the WindowsContainerSecurityContext:
-```
+Update WindowsContainerSecurityContext by adding privilged field:
+
+```protobuf
 message WindowsContainerSecurityContext {
-	string run_as_username = 1;
-	string credential_spec = 2;
+  string run_as_username = 1;
+  string credential_spec = 2;
   bool privileged = 3;
 }
 ```
-### Kubelet Implementation Details
 
-#### <b>Privileged Flag</b>
-There are no kubernetes API changes required.  Kubelet can use the privileged flag from SecurityContext api and pass that to the new privileged flag in the CRI layer.  
+*Note:* For alpha annotaions on RunPodSandbox and CreateContainer CRI calls may be used untill a version of containerd with Windows privileged container support is released.
+
+### Kubernetes API updates
+
+#### Privilged Flag
+
+A new boolean field named `privileged` will be added to [WindowsSecurityContextOptions](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#windowssecuritycontextoptions-v1-core).
+
+On Windows, all containers in a pod mush be privileged. Because of this behavior and because `WindowsSecurityContextOptions` already exists on both `PodSecurityContext` and `Container.SecurityContext` Windows containers will use this new field instead of re-using the existing `privileged` field which only exists on `SecurityContext`.
```

Comment:
Adding the field to `WindowsSecurityContextOptions` so it is set at the pod level could be ok, but I would recommend requiring (in validation) that pods that set `spec.securityContext.windowsOptions.privileged=true` also set `securityContext.privileged=true` on all containers. Policy tools already look at the container field... letting that be `false` while adding another field that makes a pod privileged will confuse them.

Reply:
Would like like to see us enforce through validation that all containers set

`container.securityContext.privilege = true`

``` yaml
spec:
  containers:
  - name: foo
    securityContext:
      privileged: true
```

`contianer.securityContext.windowsOptions.privileged = true`

``` yaml
spec:
  containers:
  - name: foo
    securityContext:
      windowsOptions:
        privileged: true
```

or both?

``` yaml
spec:
  containers:
  - name: foo
    securityContext:
      privileged: true
      windowsOptions:
        privileged: true
```

It we only require/validate `windowsOptions.privileged = true` is set on all containers (in addition to the pod-level field) would that be sufficient for policy tools?

Reply:
> It we only require/validate `windowsOptions.privileged = true` is set on all containers (in addition to the pod-level field) would that be sufficient for policy tools?

no existing policy tools would know about the new field... it seems misleading to allow privileged containers that don't set the existing privileged field in the API

Reply:
@liggitt  If we require that `securityContext.privileged=true` is set for all containers what are your thoughts on the current behavior where pod-wide WindowsSecurityContextOptions are applied to all containers if not present?

Would it be ok if the pod-wide `WindowsSecuriyContextOptions.privileged=true` and each container only sets `securityOptions.privileged=true`

for example would this be OK

``` yaml
spec:
  securityContext:
    windowsOptions:
      privileged: true
  containers:
  - name: foo
    securityContext:
      privileged: true
```

 or if the pod-wide privileged field is true ensure each container sets `securityOptions.privileged=true` and also explicitly sets `securityContext.windowsOptions.privlieged`?
 
 example:

```yaml
spec:
  securityContext:
    windowsOptions:
      privileged: true
  containers:
  - name: foo
    securityContext:
      privileged: true
      windowsOptions:
        privileged: true
```

Reply:
As discussed below, both myself and random-liu feel it would be a better user experience if we didn't rely on existing securityContext.privileged field.
Hopefully with sufficient documentation and announcement policy tools can learn about the new windowsOption.hostProcess flag for Windows containers.

Reply:
Yeah, some more context in https://github.com/kubernetes/enhancements/pull/2288#discussion_r565283865